### PR TITLE
feat(start.ps1): expand hermes-agent candidate paths for Windows installers

### PR DESCRIPTION
**Pull request by Koraji95-coder**
> **Update notes (2026-05-23, post-initial-filing):**
>
> - **PR body reformatted to match `CONTRIBUTING.md`.** Initial filing used a generic `Summary / What changed / Test plan` layout; the maintainer's `CONTRIBUTING.md` requires the 6-section shape below (Thinking Path / What Changed / Why It Matters / Verification / Risks / Model Used) plus an AI Usage Disclosure. Caught + corrected after human-side review of upstream contributing guidance.
> - **Copilot review feedback addressed in commit `6f42353`.** Three of Copilot's six inline comments were real bugs and have been fixed in this PR (`Test-Path -PathType Container` × 2 sites, conditional candidate-list build to guard `${env:ProgramFiles(x86)}` null-on-32-bit, CHANGELOG entry rewritten to be stack-position-independent). Two WSL2-related comments (`start.ps1:16` header, `README.md:140`) are out-of-scope for this PR — the affected lines originated in parent PR #2783, not #2805 — and have been replied-to on-thread with a deferral note + commitment to file a separate small docs-fix PR per the maintainer's "keep diffs focused" guidance. One comment about the CHANGELOG accuracy was rolled into the rewrite above.

---

## Thinking Path

- Hermes WebUI aims for frictionless setup on every supported platform
- PR #2783 added `start.ps1` for native Windows, mirroring `start.sh`'s discovery shape
- That discovery list only checked two paths: `%USERPROFILE%\.hermes\hermes-agent` (user-local install) and `../hermes-agent` (dev-checkout layout)
- Users hitting `start.ps1` from a fresh checkout where hermes-agent was installed via the official Windows installer (which places it under `%LOCALAPPDATA%\hermes\hermes-agent`) or via MSI (typically `Program Files`) currently have to set `HERMES_WEBUI_AGENT_DIR` manually
- That's exactly the friction PR #2783's `start.ps1` was trying to remove
- The maintainer surfaced this in the [PR #2783 review thread](https://github.com/nesquena/hermes-webui/pull/2783#issuecomment-4525735230) and explicitly suggested filing as a separate small PR per "keeping diffs focused is good policy here"

## What Changed

`start.ps1` candidate list expands from 2 → 5 paths, built incrementally with null-guards:

| Order | Path | Use case | Guard |
|---|---|---|---|
| 1 | `%USERPROFILE%\.hermes\hermes-agent` | User-local install | Always-set on Windows; unguarded |
| 2 | `%LOCALAPPDATA%\hermes\hermes-agent` | Official Windows installer | `if ($root)` |
| 3 | `%PROGRAMFILES%\hermes\hermes-agent` | MSI to Program Files | `if ($root)` |
| 4 | `%PROGRAMFILES(X86)%\hermes\hermes-agent` | MSI to x86 path | `if ($root)` — null on 32-bit Windows |
| 5 | `../hermes-agent` | Dev-checkout layout | Path-derived from `$RepoRoot`; unguarded |

Each candidate is validated as a **directory** via `Test-Path … -PathType Container` — a file (not a directory) named `hermes_cli` at any candidate root no longer passes the check. Applied uniformly at both the env-var-override-validation site (`start.ps1:94`) and the candidate-loop site (`start.ps1:111`).

Not-found error now enumerates every searched path via `$candidates -join ', '`, so when discovery fails the user sees the exact list checked.

Diff: `start.ps1 | 14 +++++++-----`, `CHANGELOG.md | 1 +` (entry under existing `[Unreleased]` / `### Added`).

## Why It Matters

The official Windows installer for hermes-agent ships it to `%LOCALAPPDATA%\hermes\hermes-agent`. Without this PR, every user of that installer who clones hermes-webui and runs `start.ps1` has to either:

- Read the not-found error, figure out which env var to set, set `HERMES_WEBUI_AGENT_DIR`, re-run
- OR move/symlink the agent install into `%USERPROFILE%\.hermes\hermes-agent`

Adding the three system-wide candidate paths means installer-deployed agents auto-discover; restructuring the error message means when discovery fails on a machine with no agent installed, the user sees all paths that were checked. The conditional-build hardening means the launcher degrades gracefully on 32-bit Windows where `${env:ProgramFiles(x86)}` is null and `Join-Path` would otherwise throw. The `-PathType Container` validation means a stray file named `hermes_cli` can't pass the discovery check and lead to a confusing import error later.

## Verification

- [x] `start.ps1` parses clean via `[System.Management.Automation.Language.Parser]::ParseFile($path, [ref]$null, [ref]$errs)` — zero errors
- [x] Anchor-based edits (no line-number drift if surrounding code shifts later); all three anchors match exactly once in the source — no false-positive replacements
- [x] `CHANGELOG.md` entry under `[Unreleased]` / `### Added`, immediately above the existing PR #2783 entry, preserves chronological ordering
- [x] Wording of CHANGELOG entry is stack-position-independent — reads correctly both when viewed in this PR's stacked diff (where `start.ps1` is "newly added") AND post-#2783-merge (where `start.ps1` is "extended")
- [ ] **Needs Windows test machine**: smoke test on a fresh checkout, hermes-agent installed at `%LOCALAPPDATA%\hermes\hermes-agent`, no `HERMES_WEBUI_AGENT_DIR` set — expected: discovery finds the agent and `start.ps1` launches `server.py`. Happy to coordinate with anyone who has the installer-deployed agent layout available; can run on a fresh foundry-side checkout if useful — let me know if you'd like before/after screenshots in a follow-up comment.
- [ ] **Needs Windows test machine**: smoke test on a fresh checkout, hermes-agent NOT installed anywhere — expected: error message lists all 5 candidate paths (or 3-4 on 32-bit / restricted-env Windows where some system-wide roots are null).
- [ ] **Needs 32-bit Windows or constrained env**: confirm conditional `if ($root)` correctly skips the missing `${env:ProgramFiles(x86)}` without `Join-Path` throwing. Same coordination offer.

Repo doesn't have PowerShell-specific tests (test suite is pytest-based for the Python side), so the PowerShell-only changes here can't be CI-tested in the existing `tests/` layout. If you'd like a Pester test added alongside this PR (or a follow-up), say the word.

## Risks / Follow-ups

- **No breaking changes.** User-local and dev-checkout candidate positions are preserved (slots 1 and 5). New candidates slot in between them. The `-PathType Container` tightening only *narrows* what passes — anyone whose `hermes_cli` was previously a file (impossible by design but theoretically possible) would now be caught earlier instead of later.
- **No new dependencies, no build-step changes, no new infrastructure.** Pure addition to an existing PowerShell launcher.
- **Stacking on #2783.** `start.ps1` doesn't exist on `master` yet; this PR depends on #2783 landing first. Once #2783 merges, rebase is clean (additive changes only).
- **Follow-up PR coming** for the WSL2-once-then-native claim in `start.ps1:16` header and `README.md:140` — Copilot flagged it as misleading (a WSL2-created venv is Linux-only, not portable to Windows Python). Out of this PR's scope; will file a separate small docs-fix PR.
- **No security-sensitive surface touched.** The discovery list just controls which directories `Test-Path` is called against; no auth, path-handling, uploads, or streaming behavior changes.

## Model Used

- **Provider:** Anthropic
- **Exact model:** Claude 4.7 (Sonnet, 1M context)
- **Mode:** Used to author the candidate-paths expansion (anchor-based PowerShell edits), the CHANGELOG entry, Copilot-fix follow-up commit, and this PR description. Anchor-based edits were verified via `[regex]::Escape($anchor)` match-once-or-fail before applying. Final human review (catching the initial PR-body format mismatch with `CONTRIBUTING.md`) before push.

### CODE REVIEW

File: start.ps1 [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    Mirrors start.sh's discovery: load optional .env, find Python,
+    locate the hermes-agent install, set sensible env defaults, then
+    invoke server.py directly. The bootstrap.py path is skipped
+    because it currently raises on platform.system() == 'Windows';
+    server.py itself runs cleanly on native Windows.
+
+    Assumes Python + hermes-agent + the WebUI Python deps are already
+    installed - same assumption start.sh makes when invoked outside
+    a fresh bootstrap. For first-time setup, run bootstrap.py inside
+    WSL2 once to create the venv, then this script can use that venv.
```
[the remark this answers is not in the document]
Comment (Koraji95-coder) [reply start.ps1:16]: Out-of-scope for #2805 — the WSL2-once-then-native claim lives in the start.ps1 header comment from PR #2783 (the parent this stacks on). #2805 only touches the candidate-path discovery list + the not-found error message. You're right the claim is technically incorrect (a WSL2-created venv is Linux-only, not portable to Windows Python), and I'll file a separate small docs-fix PR to address it cleanly per the maintainer's 'keep diffs focused' guidance.

File: README.md [hunk 2]
```diff
@@ -131,7 +131,13 @@ The bootstrap will:
...
-> A community-maintained native Windows guide is tracked in [#1952](https://github.com/nesquena/hermes-webui/issues/1952).
+
+A community-maintained native Windows setup is documented at [@markwang2658/hermes-windows-native-guide](https://github.com/markwang2658/hermes-windows-native-guide) (companion setup repo: [@markwang2658/hermes-windows-native](https://github.com/markwang2658/hermes-windows-native)). Notes from the community report in [#1952](https://github.com/nesquena/hermes-webui/issues/1952):
+
+- **Memory:** community-measured ~330 MB native vs ~1080 MB with WSL2+Docker (varies by configuration).
+- **What works:** chat, workspace browser, session management, all themes.
+- **Known limitations:** some POSIX-style file paths surface in the workspace browser; bash-assuming agent tools may not work natively.
+- **WSL2 relationship:** WSL2 is recommended *once* for first-time venv creation (since `bootstrap.py` currently refuses on native Windows). After the venv exists, `start.ps1` at the repo root runs the WebUI natively by invoking `server.py` directly — no WSL2 needed for day-to-day use.
```
[the remark this answers is not in the document]
Comment (Koraji95-coder) [reply README.md:140]: Same as the start.ps1:16 thread — this README claim originated in #2783, not #2805. Will fix in a separate small docs-fix PR rather than expanding this PR's scope.

**Comment by Koraji95-coder** [comment]
## Smoke-tested on a Windows 11 Pro machine with hermes-agent installed at `%LOCALAPPDATA%\hermes\hermes-agent`

Fresh `gh repo clone` of this PR branch, two tests + a real WOW64 edge case found.

### Test 1 — discovery via the new `%LOCALAPPDATA%` candidate (with `HERMES_WEBUI_AGENT_DIR` unset)

```
Pre-test state:
  HERMES_WEBUI_AGENT_DIR: (unset, as expected for Test 1)
  %USERPROFILE%\.hermes\hermes-agent\hermes_cli  exists-as-dir? False
  %LOCALAPPDATA%\hermes\hermes-agent\hermes_cli  exists-as-dir? True

Candidate list (5 entries, post-WOW64 fix):
  [1] C:\Users\koraj\.hermes\hermes-agent              -PathType Container hit: False
  [2] C:\Users\koraj\AppData\Local\hermes\hermes-agent -PathType Container hit: True  ← discovers here
  [3] C:\Program Files\hermes\hermes-agent             -PathType Container hit: False
  [4] C:\Program Files (x86)\hermes\hermes-agent       -PathType Container hit: False
  [5] C:\Users\koraj\AppData\Local\Temp\hermes-agent   -PathType Container hit: False

RESULT: DISCOVERY SUCCESS
  AgentDir resolved to: C:\Users\koraj\AppData\Local\hermes\hermes-agent
```

### Test 2 — `-PathType Container` guard with `HERMES_WEBUI_AGENT_DIR` set to a missing path

```
HERMES_WEBUI_AGENT_DIR = C:\does-not-exist-smoke-test-only
Test-Path -PathType Container: False

Captured stderr (process exited with code 1):
  Write-Error: ...\start.ps1:95
  HERMES_WEBUI_AGENT_DIR is set to 'C:\does-not-exist-smoke-test-only' but no hermes_cli/ folder exists there.
  Unset the variable to fall back to auto-discovery, or fix the path.
```

✓ Both new behaviors land correctly: `-PathType Container` validation fires immediately on a missing path; the override path stays a hard `exit 1` with the actionable error message.

---

### Real bug surfaced by the smoke test → fixed in commit `dbebbed`

The smoke test was running under PowerShell 7.5.4 in **WOW64 mode** (32-bit process on 64-bit OS — happens through certain shell chains, like Git Bash → pwsh). In that mode, Windows redirects `$env:ProgramFiles` to `C:\Program Files (x86)`, which meant the original candidate list produced **two identical `C:\Program Files (x86)\hermes\hermes-agent` entries and silently missed the real `C:\Program Files\hermes\hermes-agent`** where MSI-installed hermes-agent actually lives.

Initial candidate list (pre-fix, on WOW64):

```
  [3] C:\Program Files (x86)\hermes\hermes-agent   ← ${env:ProgramFiles} redirected!
  [4] C:\Program Files (x86)\hermes\hermes-agent   ← ${env:ProgramFiles(x86)} same path
```

Fix in `dbebbed`:
1. Added `${env:ProgramW6432}` to the candidate loop — Windows guarantees this points at the 64-bit `C:\Program Files` regardless of process bitness.
2. Added `$candidates = $candidates | Select-Object -Unique` after the list is built — collapses any same-path collisions defensively (handles future env-var weirdness too).

Post-fix candidate list on the same WOW64 shell — now distinct, 5 entries, includes the real `C:\Program Files\hermes\hermes-agent`.

---

This is the "Windows test machine, no `HERMES_WEBUI_AGENT_DIR` set, hermes-agent installed at `%LOCALAPPDATA%`" verification checkbox from the PR body. Calling that one ✓ and adding the WOW64 fix to the verification list.

If you'd like before/after Lighthouse-style screenshots in addition to this transcript, happy to capture them via Chrome MCP and post.

File: start.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (Koraji95-coder) [reply start.ps1:16]: Filed the separate docs-fix PR per the deferral plan: nesquena/hermes-webui#2806. Stacks on #2783, rewrites both this spot and the corresponding sister-claim to describe the native Windows venv path (Python 3.11+ install + python -m venv venv + pip install from PowerShell) and to note WSL2 venvs are not portable to Windows Python. Both #2805 WSL2-related threads now resolved with feedback flowing to #2806.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Koraji95-coder) [reply README.md:140]: Filed the separate docs-fix PR per the deferral plan: nesquena/hermes-webui#2806. Stacks on #2783, rewrites both this spot and the corresponding sister-claim to describe the native Windows venv path (Python 3.11+ install + python -m venv venv + pip install from PowerShell) and to note WSL2 venvs are not portable to Windows Python. Both #2805 WSL2-related threads now resolved with feedback flowing to #2806.

File: start.ps1 [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    exit 1
+}
+if (-not $AgentDir) {
+    $candidates = @(
+        (Join-Path $env:USERPROFILE '.hermes\hermes-agent'),
+        (Join-Path $env:LOCALAPPDATA 'hermes\hermes-agent'),
+        (Join-Path ${env:ProgramFiles} 'hermes\hermes-agent'),
+        (Join-Path ${env:ProgramFiles(x86)} 'hermes\hermes-agent'),
+        (Join-Path (Split-Path -Parent $RepoRoot) 'hermes-agent')
+    )
```
[the remark this answers is not in the document]
Comment (Koraji95-coder) [reply start.ps1]: Fixed in commit `6f42353` (initial Join-Path null-guard via conditional candidate-list build) + commit `dbebbed` (follow-up WOW64 fix that adds `${env:ProgramW6432}` and `Select-Object -Unique` to also handle 32-bit-on-64-bit PowerShell processes where `$env:ProgramFiles` gets redirected). Verified end-to-end against a WOW64 pwsh 7.5.4 — see the [smoke-test transcript comment](https://github.com/nesquena/hermes-webui/pull/2805#issuecomment-4526555722). Resolving thread.

File: start.ps1 [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+# === Find Hermes Agent dir (server.py imports from it) =================
+# When HERMES_WEBUI_AGENT_DIR is set we still validate it on disk —
+# an explicit override pointing at a missing dir should fail FAST
+# with a clear message, not silently progress into a python3 launch
+# that's about to crash on missing imports. Smoke-test feedback on
+# PR #2783: nesquena/hermes-webui requested this guard.
+$AgentDir = $env:HERMES_WEBUI_AGENT_DIR
+if ($AgentDir -and -not (Test-Path (Join-Path $AgentDir 'hermes_cli'))) {
```
[the remark this answers is not in the document]
Comment (Koraji95-coder) [reply start.ps1]: Fixed in commit `6f42353` — both `Test-Path` sites (env-var-validation at L94, candidate-loop at L111) now use `-PathType Container`. A file named `hermes_cli` at any candidate root can no longer pass the discovery guard. Resolving thread.

File: start.ps1 [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+if (-not $AgentDir) {
+    $candidates = @(
+        (Join-Path $env:USERPROFILE '.hermes\hermes-agent'),
+        (Join-Path $env:LOCALAPPDATA 'hermes\hermes-agent'),
+        (Join-Path ${env:ProgramFiles} 'hermes\hermes-agent'),
+        (Join-Path ${env:ProgramFiles(x86)} 'hermes\hermes-agent'),
+        (Join-Path (Split-Path -Parent $RepoRoot) 'hermes-agent')
+    )
+    foreach ($c in $candidates) {
+        if (Test-Path (Join-Path $c 'hermes_cli')) { $AgentDir = $c; break }
```
[the remark this answers is not in the document]
Comment (Koraji95-coder) [reply start.ps1]: Fixed in commit `6f42353` (same fix as the sister thread at L94) — `-PathType Container` now applies uniformly at both sites. Resolving thread.

File: CHANGELOG.md [hunk 6]
```diff
@@ -3,6 +3,15 @@
...
+- **`start.ps1`: native Windows launcher** — PowerShell equivalent of `start.sh` that bypasses `bootstrap.py`'s `ensure_supported_platform()` refusal and invokes `server.py` directly on native Windows. Mirrors `start.sh`'s discovery (load optional `.env` with the same readonly-var filter for `UID`/`GID`/`EUID`/`EGID`/`PPID`, find Python via `HERMES_WEBUI_PYTHON` env → `python3` → `python` → `py`, locate the hermes-agent dir at `%USERPROFILE%\.hermes\hermes-agent` or `../hermes-agent`, prefer the agent's `venv\Scripts\python.exe` if present, set `HERMES_WEBUI_HOST` / `HERMES_WEBUI_PORT` / `HERMES_WEBUI_STATE_DIR` / `HERMES_HOME` defaults). Closes the launcher half of #1952; complements the README community-guide link below. Assumes Python + agent venv are already set up — for first-time setup, use WSL2 once to create the venv, then `start.ps1` works natively.
```
[the remark this answers is not in the document]
Comment (Koraji95-coder) [reply CHANGELOG.md]: CHANGELOG entry rewritten in commit `6f42353` to be stack-position-independent — describes the discovery-path expansion without claiming start.ps1 is either newly added or pre-existing, and now includes the `-PathType Container` validation + conditional candidate-build hardening. Resolving thread.

**Comment by nesquena-hermes** [comment]
## Summary

Read the discovery block at the PR's HEAD (commit dbebbeddf) and walked through the smoke-test transcript above. The WOW64 finding is a real bug that would have shipped silently; the `ProgramW6432` + `Select-Object -Unique` fix is the right shape. Worth flagging this as the highest-value piece of the whole stack: the surface change looks small but it prevents a class of "agent installed but not found" reports from MSI-deployed-to-Program-Files users running PowerShell from inside Git Bash, VS Code's integrated terminal under certain configs, or anything else that ends up in a WOW64 process.

## Code reference

`start.ps1:104-122` at HEAD - the post-fix candidate loop:

```powershell
$candidates = @()
$candidates += (Join-Path $env:USERPROFILE '.hermes\hermes-agent')
foreach ($root in @($env:LOCALAPPDATA, ${env:ProgramW6432}, ${env:ProgramFiles}, ${env:ProgramFiles(x86)})) {
    if ($root) { $candidates += (Join-Path $root 'hermes\hermes-agent') }
}
$candidates += (Join-Path (Split-Path -Parent $RepoRoot) 'hermes-agent')
# De-dup: when running in a WOW64 (32-bit-on-64-bit) PowerShell process,
# $env:ProgramFiles is redirected to C:\Program Files (x86)...
$candidates = $candidates | Select-Object -Unique
foreach ($c in $candidates) {
    if (Test-Path (Join-Path $c 'hermes_cli') -PathType Container) { $AgentDir = $c; break }
}
```

## Diagnosis

**WOW64 bug + fix:** correct and well-targeted. Windows' filesystem-and-environment redirection layer (WoW64) maps `%ProgramFiles%` to the x86 path for 32-bit processes on 64-bit OS. The canonical escape hatch is `%ProgramW6432%`, which always points at the native 64-bit `C:\Program Files` regardless of process bitness. Without it, a 32-bit PowerShell process (which is more common than people expect - some VS Code terminal configs, Git Bash -> pwsh chains, older shell shortcuts) would have produced exactly the symptom in the transcript: `%ProgramFiles%` and `%ProgramFiles(x86)%` resolve to the same path, the real `C:\Program Files\hermes\hermes-agent` is silently skipped, and a working installer-deployed agent shows up as "not found." The `Select-Object -Unique` post-pass is also good defensive practice - it collapses any future env-var collisions before they become a "we checked twice but missed once" bug. LGTM.

**`-PathType Container` tightening:** also correct. The earlier `Test-Path (Join-Path $c 'hermes_cli')` would have passed if there was a stray *file* named `hermes_cli` at any candidate root. That's a contrived edge case but the tightening costs nothing and produces a clearer failure-point (no agent vs malformed agent) when discovery falls through. Applied uniformly at both the env-var-override-validation site (line ~95) and the candidate-loop site (line ~120). LGTM.

**Enumerated not-found error:** `Write-Error "hermes-agent not found. Searched: $searched. ..."` is a meaningful improvement over the previous "expected $primary or $sibling" wording - when discovery fails on a machine where the user thinks they installed the agent, the error names every directory checked, which makes the diagnostic loop one round instead of three.

## Smoke-test transcript

The transcript above (Test 1 confirming `%LOCALAPPDATA%` discovery, Test 2 confirming `-PathType Container` validation, plus the WOW64 finding) is the right shape of evidence. The two ticked checkboxes from the PR body are satisfied. The remaining "needs 32-bit Windows or constrained env" item could be done virtually via `Start-Job -InitializationScript { $env:'ProgramFiles(x86)' = $null }` for a unit-style proof, but it's not blocking - the conditional `if ($root)` is verifiable by inspection.

## Stacking note

Stacked on #2783 alongside #2806 and #2807 - this is the most substantive of the three follow-ups (real bug fix surfaced by smoke testing, not just hardening or docs). Worth landing.

## Verification

I did not execute the script (cron policy is read-only on PR worktrees). The WOW64 redirect behavior is documented Windows API surface; the fix is verifiable by inspection. The `-PathType Container` change is mechanically equivalent at the happy-path level but tightens the failure mode. The enumerated error message is structurally observable in the diff.

**Comment by nesquena-hermes** [comment]
## Summary

Read the discovery block at the PR's HEAD (commit dbebbeddf) and walked through the smoke-test transcript above. The WOW64 finding is a real bug that would have shipped silently; the `ProgramW6432` + `Select-Object -Unique` fix is the right shape. Worth flagging this as the highest-value piece of the whole stack: the surface change looks small but it prevents a class of "agent installed but not found" reports from MSI-deployed-to-Program-Files users running PowerShell from inside Git Bash, VS Code's integrated terminal under certain configs, or anything else that ends up in a WOW64 process.

## Code reference

`start.ps1:104-122` at HEAD - the post-fix candidate loop:

```powershell
$candidates = @()
$candidates += (Join-Path $env:USERPROFILE '.hermes\hermes-agent')
foreach ($root in @($env:LOCALAPPDATA, ${env:ProgramW6432}, ${env:ProgramFiles}, ${env:ProgramFiles(x86)})) {
    if ($root) { $candidates += (Join-Path $root 'hermes\hermes-agent') }
}
$candidates += (Join-Path (Split-Path -Parent $RepoRoot) 'hermes-agent')
# De-dup: when running in a WOW64 (32-bit-on-64-bit) PowerShell process,
# $env:ProgramFiles is redirected to C:\Program Files (x86)...
$candidates = $candidates | Select-Object -Unique
foreach ($c in $candidates) {
    if (Test-Path (Join-Path $c 'hermes_cli') -PathType Container) { $AgentDir = $c; break }
}
```

## Diagnosis

**WOW64 bug + fix:** correct and well-targeted. Windows' filesystem-and-environment redirection layer (WoW64) maps `%ProgramFiles%` to the x86 path for 32-bit processes on 64-bit OS. The canonical escape hatch is `%ProgramW6432%`, which always points at the native 64-bit `C:\Program Files` regardless of process bitness. Without it, a 32-bit PowerShell process (which is more common than people expect - some VS Code terminal configs, Git Bash -> pwsh chains, older shell shortcuts) would have produced exactly the symptom in the transcript: `%ProgramFiles%` and `%ProgramFiles(x86)%` resolve to the same path, the real `C:\Program Files\hermes\hermes-agent` is silently skipped, and a working installer-deployed agent shows up as "not found." The `Select-Object -Unique` post-pass is also good defensive practice - it collapses any future env-var collisions before they become a "we checked twice but missed once" bug. LGTM.

**`-PathType Container` tightening:** also correct. The earlier `Test-Path (Join-Path $c 'hermes_cli')` would have passed if there was a stray *file* named `hermes_cli` at any candidate root. That's a contrived edge case but the tightening costs nothing and produces a clearer failure-point (no agent vs malformed agent) when discovery falls through. Applied uniformly at both the env-var-override-validation site (line ~95) and the candidate-loop site (line ~120). LGTM.

**Enumerated not-found error:** `Write-Error "hermes-agent not found. Searched: $searched. ..."` is a meaningful improvement over the previous "expected $primary or $sibling" wording - when discovery fails on a machine where the user thinks they installed the agent, the error names every directory checked, which makes the diagnostic loop one round instead of three.

## Smoke-test transcript

The transcript above (Test 1 confirming `%LOCALAPPDATA%` discovery, Test 2 confirming `-PathType Container` validation, plus the WOW64 finding) is the right shape of evidence. The two ticked checkboxes from the PR body are satisfied. The remaining "needs 32-bit Windows or constrained env" item could be done virtually via `Start-Job -InitializationScript { $env:'ProgramFiles(x86)' = $null }` for a unit-style proof, but it's not blocking - the conditional `if ($root)` is verifiable by inspection.

## Stacking note

Stacked on #2783 alongside #2806 and #2807 - this is the most substantive of the three follow-ups (real bug fix surfaced by smoke testing, not just hardening or docs). Worth landing.

## Verification

I did not execute the script (cron policy is read-only on PR worktrees). The WOW64 redirect behavior is documented Windows API surface; the fix is verifiable by inspection. The `-PathType Container` change is mechanically equivalent at the happy-path level but tightens the failure mode. The enumerated error message is structurally observable in the diff.

**Comment by nesquena-hermes** [comment]
Shipped in v0.51.124 via release/stage-batch6 (#2817). The expanded candidate-list discovery + WOW64 dedupe + PathType Container guard are in. Rebased onto current master (since this was branched off the now-shipped #2783); authorship preserved. Thanks!
